src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,344 @@
+                    }
+
+                    break;
+                }
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCnot(diag, ctrl, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+
+                    break;
+                }
+
+                case qc::OpType::I: {
+                    break;
+                }
+
+                case qc::OpType::Phase: {
+                    const auto phase = PiRational(op->getParameter().front());
+                    addCphase(diag, phase, ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::T: {
+                    addCphase(diag, PiRational(1, 4), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::S: {
+                    addCphase(diag, PiRational(1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Tdag: {
+                    addCphase(diag, PiRational(-1, 4), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Sdag: {
+                    addCphase(diag, PiRational(-1, 2), ctrl, target, qubit_vertices);
+                    break;
+                }
+
+                default: {
+                    throw ZXException("Unsupported Controlled Operation: " +
+                                      qc::toString(op->getType()));
+                }
+            }
+        } else if (op->getNcontrols() == 2) {
+            Qubit       ctrl_0 = 0;
+            Qubit       ctrl_1 = 0;
+            const Qubit target = op->getTargets().front();
+            int         i      = 0;
+            for (auto& ctrl: op->getControls()) {
+                if (i++ == 0)
+                    ctrl_0 = ctrl.qubit;
+                else
+                    ctrl_1 = ctrl.qubit;
+            }
+            switch (op->getType()) {
+                case qc::OpType::X: {
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    break;
+                }
+
+                case qc::OpType::Z: {
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    addCcx(diag, ctrl_0, ctrl_1, target, qubit_vertices);
+                    addZSpider(diag, target, qubit_vertices, Expression(),
+                               EdgeType::Hadamard);
+                    break;
+                }
+                default: {
+                    throw ZXException("Unsupported Multi-control operation: " +
+                                      qc::toString(op->getType()));
+                    break;
```

Comment:
Maybe it also makes sense to test these unsupported operations.
In that fashion, once they are supported, there already is a test for that.